app/src/main/java/eu/kanade/tachiyomi/ui/reader/loader/SevenZipPageLoader.kt

```diff
@@ -0,0 +1,40 @@
+package eu.kanade.tachiyomi.ui.reader.loader
+
+import com.hippo.unifile.UniFile
+import eu.kanade.tachiyomi.source.model.Page
+import eu.kanade.tachiyomi.ui.reader.model.ReaderPage
+import eu.kanade.tachiyomi.util.storage.SevenZUtil.getImages
+import org.apache.commons.compress.archivers.sevenz.SevenZFile
+import org.apache.commons.compress.utils.SeekableInMemoryByteChannel
+
+/**
+ * Loader used to load a chapter from a .7z or .cb7 file.
+ */
+internal class SevenZipPageLoader(
+    private val file: UniFile,
+    private val notifySlowArchive: (method: String) -> Unit,
+) : PageLoader() {
+
+    private val zip by lazy { SevenZFile(SeekableInMemoryByteChannel(file.openInputStream().readBytes())) }
```

Comment:
Read the file completely isn't ideal.
Instead, you can construct a `FileInputStream` and use `FileInputStream.getChannel`.
You can take https://github.com/FooIbar/EhViewer/blob/aec34c81aa586d2edc3d2a5cf494d238ddae3001/app/src/main/kotlin/com/hippo/unifile/UniFileExtensions.kt#L14 for reference.

Reply:
Like so? I couldn't get it to work since the URI isn't of "file" scheme.
```kotlin
private val zip by lazy { SevenZFile(File(URI(file.uri.toString())).inputStream().channel) }
```

Reply:
Use https://developer.android.com/reference/android/content/ContentResolver#openFileDescriptor(android.net.Uri,%20java.lang.String) then wrap it into a `FileInputStream`.

Reply:
Doesn't the ContentResolver require a context?

Reply:
Yes, can just use application context.

Reply:
I can't access the application context inside the SevenZipPageLoader, should i just pass the one from the ChapterLoader?

Reply:
doing this causes an IO exception of "Bad file descriptor".
![image](https://github.com/mihonapp/mihon/assets/54363735/efe5ff2a-4b0b-4075-9e0b-0e278c8f0778)

Reply:
Didn't I say wrap it into AutoCloseInputStream (which extends FileInputStream)?

Reply:
sorry, didn't see that. thank you for your patience